langchain_lakefs/document_loaders.py

```diff
@@ -1,72 +1,161 @@
+            for obj in objects  # Iterate over ObjectInfo instances
+            for doc in UnstructuredLakeFSLoader(
+                obj.physical_address,  # Extract physical_address
+                self.repo,
+                self.ref,
+                obj.path,  # Extract path
+                presigned,
+                user_metadata=obj.metadata,  # Extract metadata
+                client=self.client,
+            ).load()
+        ]
+
+        return documents
+
+    def __validate_instance(self) -> None:
+        if self.repo is None or self.repo == "":
+            raise ValueError(
+                "no repository was provided. use `set_repo` to specify a repository"
+            )
+        if self.ref is None or self.ref == "":
+            raise ValueError("no ref was provided. use `set_ref` to specify a ref")
+        if self.path is None:
+            raise ValueError("no path was provided. use `set_path` to specify a path")
+
+
+class UnstructuredLakeFSLoader(UnstructuredBaseLoader):
+    """Load from `lakeFS` as unstructured data."""
+
+    def __init__(
+            self,
+            url: str,
+            repo: str,
+            ref: str = "main",
+            path: str = "",
+            presign: bool = True,
+            client: Optional[Client] = None,
+            # presign: bool = False,
+            user_metadata: Optional[dict] = None,
+            **unstructured_kwargs: Any,
+    ):
+        """Initialize UnstructuredLakeFSLoader.
+
+        Args:
+        :param url:
+        :param repo:
+        :param ref:
+        :param path:
+        :param presign:
+        :param user_metadata:
+        :param lakefs_access_key:
+        :param lakefs_secret_key:
+        :param lakefs_endpoint:
+        """
+
+        super().__init__(**unstructured_kwargs)
+        self.user_metadata = user_metadata
+        self.url = url
+        self.repo = repo
+        self.ref = ref
+        self.path = path
+        self.presign = presign
+        self.client = client
+
+    def _get_metadata(self) -> dict:
+        metadata = {"repo": self.repo, "ref": self.ref, "path": self.path}
+        if self.user_metadata:
+            for key, value in self.user_metadata.items():
+                if key not in metadata:
+                    metadata[key] = value
+        return metadata
+
+    def _get_elements(self) -> List:
+        local_prefix = "local://"
+        if self.url.startswith(local_prefix):
+            local_path = self.url[len(local_prefix):]
+            return partition(filename=local_path)
+        else:
+            with tempfile.TemporaryDirectory() as temp_dir:
+                file_path = f"{temp_dir}/{self.path.split('/')[-1]}"
+                os.makedirs(os.path.dirname(file_path), exist_ok=True)
+                obj = lakefs.repository(self.repo, client=self.client).ref(self.ref).object(self.path)
+                with open(file_path, mode="wb") as file:
+                    file.write(obj.reader().read())
```

Comment:
check if we can just pass the reader and not read the content we need to write to reduce the memory usage of this operation.